### ISSUE TITLE
Suspected bug in merge_data

### ISSUE THREAD

**Issue by zyzhang1130** [alibaba/FederatedScope#384]
**Describe the bug**

`federatedscope.core.auxiliaries.data_builder.merge_data` causes `all_data[1] == merged_data`. Furthermore, it appears the size of the  merged data is not consistent with the original dataset.

**To Reproduce**

```
from federatedscope.core.configs.config import global_cfg

cfg = global_cfg.clone()



from federatedscope.core.auxiliaries.data_builder import get_data, merge_data

cfg.data.type = 'femnist'
cfg.data.splits = [0.6, 0.2, 0.2]
cfg.data.batch_size = 10
cfg.data.subsample = 0.05
cfg.data.transform = [['ToTensor'], ['Normalize', {'mean': [0.1307], 'std': [0.3081]}]]
cfg.federate.client_num = 16

data, modified_cfg = get_data(cfg.clone())
cfg.merge_from_other_cfg(modified_cfg)

merged_data = merge_data(data,5)

train_count = 0
test_count = 0
val_count = 0
for i in range(1,5):
    train_count += len(data[i]['train'])
    test_count += len(data[i]['test'])
    val_count += len(data[i]['val'])
```
**Expected behavior**
1. `data[1] == merged_data` should be `false` (there is no obvious reason why it should be true, especially when it size of `data` remains the same).
2. `train_count`, `test_count` and `val_count`  should match `merged_data`.
**Screenshots**
![image](https://user-images.githubusercontent.com/36942574/192154715-5ec87d20-e372-42d7-a797-e8a59b5e1281.png)

![image](https://user-images.githubusercontent.com/36942574/192154687-d5f25ee2-ee22-41f1-9ab0-f29fad15cb6f.png)

**Desktop (please complete the following information):**
 - OS: Windows11

**Comment by yxdyc** [comment]
Thanks for your question! 

In our original version, we merged the data `all_data` with ids [1, ..., `merged_max_data_id`] into `merged_data`, which is initialized as a shallow copy of data of client 1. The `merge_data` is used in global training mode and works well in the situation since users usually do not need to use the original data with `all_data[x]`.

To make the merge more safe, we modified this function with a deep copy initialization (in #385 ) and now the behavior will meet your expection. Please see the following test codes and output.

```
from federatedscope.core.configs.config import global_cfg

cfg = global_cfg.clone()

from federatedscope.core.auxiliaries.data_builder import get_data, merge_data

cfg.data.type = 'femnist'
cfg.data.splits = [0.6, 0.2, 0.2]
cfg.data.batch_size = 10
cfg.data.subsample = 0.05
cfg.data.transform = [['ToTensor'],
                      ['Normalize', {'mean': [0.1307], 'std': [0.3081]}]]
cfg.federate.client_num = 16

data, modified_cfg = get_data(cfg.clone())
cfg.merge_from_other_cfg(modified_cfg)

train_count = 0
test_count = 0
val_count = 0
print("====================Before merging==================")
for i in range(1, 5):
    print(f"datasize of each client are {len(data[i]['train'].dataset)},"
          f" {len(data[i]['test'].dataset)}, "
          f"{len(data[i]['val'].dataset)}")
    train_count += len(data[i]['train'].dataset)
    test_count += len(data[i]['test'].dataset)
    val_count += len(data[i]['val'].dataset)

merged_data = merge_data(data, 5)

train_count = 0
test_count = 0
val_count = 0
print("====================After merging==================")
for i in range(1, 5):
    print(f"datasize of each client are {len(data[i]['train'].dataset)},"
          f" {len(data[i]['test'].dataset)}, "
          f"{len(data[i]['val'].dataset)}")
    train_count += len(data[i]['train'].dataset)
    test_count += len(data[i]['test'].dataset)
    val_count += len(data[i]['val'].dataset)

print(f"total size of the merged data, {len(merged_data['train'].dataset)}"
      f", {len(merged_data['test'].dataset)}"
      f", {len(merged_data['val'].dataset)}")
```

The output with this PR:
![image](https://user-images.githubusercontent.com/67475544/192192067-8ab64a90-1df9-4932-ae4a-50f1d682e78f.png)

**Comment by zyzhang1130** [comment]
Noted with thanks!

**Comment by zyzhang1130** [comment]
Hi just one more calcification: is the dataset with index `merged_max_data_id` included in the `merged_data`? Because in the old script line 763 of `federatedscope.core.auxiliaries.data_builder`, it seems that `data_id` never took on the value of `merged_max_data_id`.

**Comment by yxdyc** [comment]
> Hi just one more calcification: is the dataset with index `merged_max_data_id` included in the `merged_data`? Because in the old script line 763 of `federatedscope.core.auxiliaries.data_builder`, it seems that `data_id` never took on the value of `merged_max_data_id`.

Hi, the index `merged_max_data_id` should be included in the `merged_data` as the id 0 indicates the server. We have updated the output above and the PR accordingly, thanks again~

**Comment by xieyxclack** [comment]
This thread is closed since we have fixed it in #385 
Thanks again for your comments!

**Comment by zyzhang1130** [comment]
noted with thanks
